lib/common/schemas.c

```diff
@@ -1275,3 +1281,63 @@ cli_config_update(xmlNode **xml, int *best_version, gboolean to_logs)
     free(orig_value);
     return rc;
 }
+
+/*!
+ * \internal
+ * \brief Return a list of all schema files and any associated XSLT files
+ *        later than the given one
+ * \brief Return a list of all schema versions later than the given one
+ *
+ * \param[in] schema The schema to compare against (for example,
+ *                   "pacemaker-3.1.rng" or "pacemaker-3.1")
+ *
+ * \note The caller is responsible for freeing both the returned list and
+ *       the elements of the list
+ */
+GList *
+pcmk__schema_files_later_than(const char *schema)
+{
+    GList *lst = NULL;
+    schema_version_t ver;
+
+    if (!version_from_filename(schema, &ver)) {
+        return lst;
+    }
+
+    for (int lpc = xml_latest_schema_index(); lpc > 0; lpc--) {
+        char *s = NULL;
+
+        if (schema_cmp(ver, known_schemas[lpc].version) != -1) {
+            continue;
```

Comment:
I think we actually have to traverse the whole list of schemas everywhere now.  Everything had a built-in assumption that because we were only adding schemas once from one directory, the list we called add_schema on would already be sorted (via scandir and schema_cmp_directory).  add_schema doesn't do anything to preserve order, so adding a second set of schemas is just going to cause them to get added to the end of the list.

If that's going to be a problem, I'll need to change add_schema to keep the list sorted.

Reply:
Yeah, I suspect we'll have unexpected problems if we don't keep it sorted

Reply:
The thing that's tripping up sorting at the moment is that each schema record holds the index of the schema record that it can be upgraded to.  We can replace the index with a pointer to get around that.  But I'm also wondering if there's any instances (besides "pacemaker-next" and "none") where we care.  At the moment, every schema simply upgrades to the next one in the list.  If we can assume that will always be the case, I can just get rid of this index/pointer entirely.

Reply:
That makes sense

Reply:
Added a commit on top to deal with this.  #3247 will need to be merged first, however.